src/Interpreters/MergeJoin.cpp

```diff
@@ -433,6 +434,32 @@ MergeJoin::MergeJoin(std::shared_ptr<TableJoin> table_join_, const Block & right
     , max_rows_in_right_block(table_join->maxRowsInRightBlock())
     , max_files_to_merge(table_join->maxFilesToMerge())
 {
+    const auto & key_names_left = table_join->keyNamesLeft();
+    const auto & key_names_right = table_join->keyNamesRight();
+
+    bool cast_allowed = JoinCommon::isCastJoinKeysAllowed(left_sample_block, right_sample_block, *table_join);
+    if (cast_allowed)
+    {
+        /// Left key columns will be converted to types from right if it isn't match
```

Comment:
Are we casting left columns for merge join, but right columns for hash join? Maybe we can always cast right, for consistency?

It would be interesting to add a test to see how the runtime switching from hash join to merge join works with casting.